src/rules.jl

```diff
@@ -1,260 +1,313 @@
     end
-    return msg
+    return false
 end
 
-function rule_c!(zxd::ZXDiagram{T,U,P}, v1::T, v2::T) where {T<:Integer, U<:Integer, P}
-    msg = check_rule_c(zxd, v1, v2)
-    if msg != true
-        error(msg)
+function rewrite!(zxd::ZXDiagram, r::Rule_pi, vs)
+    v1 = vs[1]
+    v2 = vs[2]
+    zxd.ps[v2] = -zxd.ps[v2]
+    nb = neighbors(zxd, v2, count_mul = true)
+    for v3 in nb
+        v3 != v1 && insert_spider!(zxd, v2, v3, X, phase(zxd, v1))
     end
-    rem_edge!(zxd, v1, v2)
-    nbhd2 = outneighbors(zxd, v2)
-    for v in nbhd2
-        insert_spider!(zxd, v2, v, X)
+    if neighbors(zxd, v1) != [v2]
+        add_edge!(zxd, neighbors(zxd, v1))
+        rem_spider!(zxd, v1)
     end
-    vmap = rem_spiders!(zxd, [v1, v2])
-
-    return vmap
+    rounding_phases!(zxd)
+    zxd
 end
 
-function check_rule_b(zxd::ZXDiagram{T,U,P}, v1::T, v2::T) where {T<:Integer, U<:Integer, P}
-    msg = true
-    if v1 <= nv(zxd) && v2 <= nv(zxd)
-        if (zxd.st[v1] == Z && zxd.st[v2] == X) || (zxd.st[v1] == X && zxd.st[v2] == Z)
-            if zxd.ps[v1] == zero(P) && zxd.ps[v1] == zero(P)
-                if has_edge(zxd.g, v1, v2)
-                    if !(degree(zxd.g, v1) == 3 && degree(zxd.g, v2) == 3)
-                        msg = "Spiders $(v1) or $(v2) is not of degree 3!"
-                    end
-                else
-                    msg = "Spiders $(v1) and $(v2) are not connected!"
-                end
-            else
-                msg = "Spiders $(v1) or $(v2) is not of phase 0!"
+function check_rule(zxd::ZXDiagram, r::Rule_c, vs)
+    vs ⊆ spiders(zxd) || return false
+    v1 = vs[1]
+    v2 = vs[2]
+    if spider_type(zxd, v1) == X && phase(zxd, v1) == zero(phase(zxd, v1)) && length(neighbors(zxd, v1, count_mul = true)) == 1
+        if v2 in neighbors(zxd, v1)
+            if spider_type(zxd, v2) == Z
+                return true
             end
-        else
-            msg = "Spiders ($(v1), $(v2)) are not a (Z, X) pair!"
         end
-    else
-        msg = "Spider $(v1) or $(v2) is not in this ZX-diagram!"
     end
-    return msg
+    return false
 end
 
-function rule_b!(zxd::ZXDiagram{T,U,P}, v1::T, v2::T) where {T<:Integer, U<:Integer, P}
-    msg = check_rule_b(zxd, v1, v2)
-    if msg != true
-        error(msg)
-    end
-    rem_edge!(zxd, v1, v2)
-    nbhd1 = outneighbors(zxd, v1)
-    for v in nbhd1
-        insert_spider!(zxd, v1, v, SType(1-Int(zxd.st[v1])))
+function rewrite!(zxd::ZXDiagram, r::Rule_c, vs)
+    v1 = vs[1]
+    v2 = vs[2]
+    ph = phase(zxd, v1)
+    rem_spider!(zxd, v1)
+    nb = neighbors(zxd, v2, count_mul = true)
+    for v3 in nb
+        add_spider!(zxd, X, ph, [v3])
     end
-    v1_1 = nv(zxd)
-    v1_2 = v1_1 - 1
+    rem_spider!(zxd, v2)
+    zxd
+end
 
-    nbhd2 = outneighbors(zxd, v2)
-    for v in nbhd2
-        insert_spider!(zxd, v2, v, SType(1-Int(zxd.st[v2])))
+function check_rule(zxd::ZXDiagram, r::Rule_b, vs)
```

Comment:
should `check_rule` be `ismatch` instead? not a big deal, but this would be more consistent with other packages.